syntax = "proto3";

package nuance.asr.v1beta1;

/*
Input message defining one or more recognition resources (domain LMs, wordsets, speaker profiles, builtins) to improve recognition.  Included in [RecognizeInitMessage](#recognizeinitmessage). Domain LMs and speaker profiles must be external references but wordsets may also be provided inline.
*/
message RecognitionResource {  
  oneof resource_union {
    ResourceReference external_reference = 1; // The resource is an external file. Mandatory for DLMs and speaker profiles. 
    string inline_wordset = 2;                // Inline wordset JSON resource. See [Wordsets](#wordsets) for the format. Default blank, meaning no inline wordset.
    string builtin = 3;          		      // Name of a builtin resource in the data pack. Default blank, meaning no builtins.
  } 
  oneof weight_union {
    EnumWeight weight_enum = 4;               // Keyword for weight of DLM relative to data pack. Default MEDIUM.
    uint32 weight_value = 5;                  // Weight of DLM relative to data pack as number from 0 to 1000. Default 0. 
  }  
}

/*
Input message for fetching an external recognition resource. Included in [RecognitionResource](#recognitionresource). See [Resources](#resources). For Nuance-hosted Krypton, the uri for DLMs is a URN, e.g. urn:nuance:mix/eng-USA/A2_C70/mix.asr. 
*/
message ResourceReference {
  EnumResourceType type = 1;     	// Resource type. Default UNDEFINED_RESOURCE_TYPE.
  string uri = 2;                	// Location of the resource as a URN reference. 
  oneof optional_resource_reference_max_age {
    uint32 max_age = 3;            	// See tools.ietf.org/html/rfc7234#section-5.2.1.1. Default 0.
  }
  oneof optional_resource_reference_max_stale {
    uint32 max_stale = 4;          	// See tools.ietf.org/html/rfc7234#section-5.2.1.2. Default 0.
  }
  oneof optional_resource_reference_min_fresh {
    uint32 min_fresh = 5;          	// See tools.ietf.org/html/rfc7234#section-5.2.1.3. Default 0.
  }
  string cookies = 6;            	// Cookies to transmit to the server. Default blank.
  uint32 request_timeout_ms = 7; 	// Time to wait when downloading resources. Default (0) means server default, usually 10000ms or 10 seconds.  
  EnumResourceReuse reuse = 8;      // Whether the resource will be used multiple times. Default LOW_REUSE.
}

/*
Input field defining the content type of a recognition resource: a domain LM, wordset (plain text or compiled) or speaker profile. Included in [ResourceReference](#resourcereference). See [Resources](#resources).
*/
enum EnumResourceType {
  UNDEFINED_RESOURCE_TYPE = 0;      // Resource type is not specified. Client must always specify a type.  
  WORDSET = 1;                      // Resource is a plain-text JSON wordset, provided inline or referenced as a file. 
  COMPILED_WORDSET = 2;             // Resource is a compiled wordset zip file. 
  DOMAIN_LM = 3;                    // Resource is a domain LM zip file.
  SPEAKER_PROFILE = 4;              // Resource is a speaker profile defined in a Minio endpoint.
}

/*
Input field specifying whether the domain LMs or wordset will be used for one or many recognition turns. Included in [ResourceReference](#resourcereference). 
*/
enum EnumResourceReuse {
  UNDEFINED_REUSE = 0;              // Not specified: currently defaults to LOW_REUSE.
  LOW_REUSE = 1;                    // The resource will be used for only one recognition turn.
  HIGH_REUSE = 5;                   // The resource will be used for a sequence of recognition turns.
  reserved 2,3,4;
}

/*
Input field setting the relative weight of the domain LM to the data pack, as a keyword. Included in [RecognitionResource](#recognitionresource). Wordsets and speaker profiles do not have a weight. See also weight_value to specify a numeric value.
*/
enum EnumWeight {
  DEFAULT_WEIGHT = 0; // Same effect as MEDIUM
  LOWEST = 1;         // The DLM has less effect than the data pack.
  LOW = 2;
  MEDIUM = 3;         // The DLM has roughly an equal effect compared to the data pack. 
  HIGH = 4;
  HIGHEST = 5;        // The DLM has more effect than the data pack. 
}









